appsec-kit-v7/src/main/java/com/vaadin/appsec/v7/service/NotificationInitializer.java

```diff
@@ -37,91 +36,49 @@ public class NotificationInitializer {
-        if (isDebugMode(vaadinService)) {
-            vaadinService.addSessionInitListener(e -> {
-                e.getSession().addUIProvider(new AppSecUIProvider());
-                startNotificationTask(e.getSession());
-            });
-            LOGGER.info("NotificationInitListener initialized.");
+    void serviceInit(VaadinService service) {
+        if (isDebugMode(service)) {
+            service.addSessionInitListener(this::subscribeSessionToScanEvents);
+            service.addSessionDestroyListener(this::removeSessionRegistration);
+            LOGGER.info("Subscribed to AppSec Kit scan events");
         }
     }
 
-    private static void startNotificationTask(final VaadinSession session) {
-        final String sessionId = session.getSession().getId();
-        final ScheduledFuture<?> scheduledNotificationCheck = AppSecService
-                .getInstance().getConfiguration().getTaskExecutor()
-                .scheduleAtFixedRate(() -> {
-                    if (isSessionOpen(session)) {
-                        session.access(() -> {
-                            session.getUIs().forEach(
-                                    ui -> notifyUiIfNeeded(session, ui));
-                        });
-                    } else {
-                        ScheduledFuture<?> thisCheck = scheduledChecks
-                                .get(sessionId);
-                        if (thisCheck != null) {
-                            thisCheck.cancel(true);
-                            scheduledChecks.remove(sessionId);
-                        }
-                    }
-                }, NOTIFICATION_CHECK_INTERVAL, NOTIFICATION_CHECK_INTERVAL,
-                        TimeUnit.MILLISECONDS);
-        scheduledChecks.put(sessionId, scheduledNotificationCheck);
-        LOGGER.info(
-                "NotificationInitListener notification thread initialized.");
+    private void removeSessionRegistration(SessionDestroyEvent e) {
+        VaadinSession session = e.getSession();
+        Registration registration = scanEventRegistrations.get(session);
+        if (registration != null) {
+            registration.remove();
+        }
+    }
 
+    private void subscribeSessionToScanEvents(SessionInitEvent e) {
+        VaadinSession session = e.getSession();
+        session.addUIProvider(new AppSecUIProvider());
+        AppSecService appSecService = AppSecService.getInstance();
+        Registration scanEventRegistration = appSecService
+                .addScanEventListener(scanEvent -> {
+                    int newVulns = scanEvent.getNewVulnerabilities().size();
+                    if (isSessionOpen(session) && newVulns > 0) {
+                        session.getUIs().forEach(this::doNotifyUI);
+                    }
+                });
+        scanEventRegistrations.put(session, scanEventRegistration);
     }
 
-    @SuppressWarnings("unchecked")
-    private static void notifyUiIfNeeded(VaadinSession session, UI ui) {
```

Comment:
If the automatic scan interval, which is configurable, is set to a short interval (e.g. couple of minutes) then without tracking that which UI has been notified already the notification popup can appear many times on one UI which covers each other and creates an increasing black border around the popup during the time.

Reply:
Even if the scan happens every 2 minutes, and new vulnerabilities are found, we still want to notify all the UIs.

Reply:
Notifying all the UIs works as expected. If I have for example three sessions opened all of them get notified at the same time, this is ok. I rather wanted to point on that these sessions keep getting notified in every 2 minutes, the old popup remains there and when the new popup arrives it covers the old one. After a while with many popups there can have a bad look and it is hard to close many popups. Even if the popup is clicked to open the AppSec UI only the last one is closed. This is when more then 20 popups are there at the same time:
![Screenshot 2023-07-04 at 17 09 55](https://github.com/vaadin/appsec-kit/assets/11977763/ddc9b730-df04-4ff9-a7d7-990e41e5b294)

Reply:
This is true, good point. It happens because we set a 24 hours delay for the notification before disappearing, in practice it never disappears until you interact with it. To solve this, we can set the delay the same as the periodic interval between automatic scans so that the notification disappears right before a new one might popup.

Reply:
Yes, this is good idea and would be a nice solution.